lib/buffer.js

```diff
@@ -150,25 +149,16 @@ function fromObject(obj) {
     return binding.createFromArrayBuffer(obj);
   }
 
-  if (obj.buffer instanceof ArrayBuffer || obj.length) {
-    let length;
-    if (typeof obj.length !== 'number' || obj.length !== obj.length)
-      length = 0;
-    else
-      length = obj.length;
-    const b = allocate(length);
-    for (let i = 0; i < length; i++) {
-      b[i] = obj[i] & 255;
+  if (obj.buffer instanceof ArrayBuffer || 'length' in obj) {
+    if (typeof obj.length !== 'number' || obj.length !== obj.length) {
+      return allocate(0);
+    } else {
+      return fromArrayLike(obj);
```

Comment:
just an aside, all typed arrays could be more quickly handled in C++, but that doesn't concern this PR. at least now it's centralized and will be easier to make the change. :)